### ISSUE TITLE
Cookbook upload fails with: OpenSSL::OpenSSLError: password must be at least 4 bytes

### ISSUE THREAD

**Issue by SteFletcher** [chef/chef#7383]
## Description
"OpenSSL::OpenSSLError: password must be at least 4 bytes" when attempting to upload cookbooks to OpsWorks.  

Initially I wasn't sure why the chef-dk was using the RVM version of ruby as the official documentation says ruby is embedded and should be used from /opt/chefdk

So I checked my environment variables and discovered lots of overrides for RUBY_HOME and also PATH (pointing to RVM's ruby gem bin folders).  I removed this from the env and everything started to work.

In the case where we use RVM, why would the hack below solve the problem?

My hack here /Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http/authenticator.rb:95

```
# Pass in '' as the passphrase to avoid OpenSSL prompting on the TTY if
# given an encrypted key. This also helps if using a single file for
# both the public and private key with ssh-agent mode.
# @key = OpenSSL::PKey::RSA.new(@raw_key, "")
@key = OpenSSL::PKey::RSA.new(@raw_key, '1234')
```

## Chef Version
Chef Development Kit Version: 3.0.44
chef-client version: 14.2.0
delivery version: master (cc1ff71724d5c36bc4bbf187a95cec4bbe81ed9c)
berks version: 7.0.4
kitchen version: 1.21.2
inspec version: 2.1.72

## Platform Version
macOS v10.13.5

## Replication Case
Use RVM to install ruby for macOS
Try to upload a cookbook to a chef server with PEM authentication (AWS OpsWorks in my case).

## Stacktrace

```
SSL_CERT_FILE='.chef/ca_certs/opsworks-cm-ca-2016-root.pem' berks upload
E, [2018-06-20T17:14:37.585039 #80087] ERROR -- : OpenSSL::OpenSSLError: password must be at least 4 bytes
E, [2018-06-20T17:14:37.585172 #80087] ERROR -- : /Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http/authenticator.rb:95:in `initialize'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http/authenticator.rb:95:in `new'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http/authenticator.rb:95:in `load_signing_key'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http/authenticator.rb:42:in `initialize'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http.rb:98:in `new'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http.rb:98:in `block in initialize'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http.rb:97:in `each'
/Users/me/.rvm/gems/ruby-2.4.1/gems/chef-14.2.0/lib/chef/http.rb:97:in `initialize'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/ridley_compat.rb:26:in `initialize'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/ridley_compat.rb:53:in `new'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/ridley_compat.rb:53:in `new_client'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf.rb:172:in `ridley_connection'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/uploader.rb:55:in `upload'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/uploader.rb:42:in `run'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/berksfile.rb:576:in `upload'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/cli.rb:195:in `upload'
/Users/me/.rvm/gems/ruby-2.4.1/gems/thor-0.20.0/lib/thor/command.rb:27:in `run'
/Users/me/.rvm/gems/ruby-2.4.1/gems/thor-0.20.0/lib/thor/invocation.rb:126:in `invoke_command'
/Users/me/.rvm/gems/ruby-2.4.1/gems/thor-0.20.0/lib/thor.rb:387:in `dispatch'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/cli.rb:47:in `dispatch'
/Users/me/.rvm/gems/ruby-2.4.1/gems/thor-0.20.0/lib/thor/base.rb:466:in `start'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/lib/berkshelf/cli.rb:23:in `execute!'
/Users/me/.rvm/gems/ruby-2.4.1/gems/berkshelf-7.0.3/bin/berks:5:in `<top (required)>'
/Users/me/.rvm/gems/ruby-2.4.1/bin/berks:23:in `load'
/Users/me/.rvm/gems/ruby-2.4.1/bin/berks:23:in `<main>'
/Users/me/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `eval'
/Users/me/.rvm/gems/ruby-2.4.1/bin/ruby_executable_hooks:15:in `<main>'
```

**Comment by SteFletcher** [comment]
Posting on stackoverflow instead, not sure this is a bug with the chef dk or something else.

**Comment by predatorian3** [comment]
I'm also getting this with Chef v14.2 when attempting to execute a `knife exec` script doing a search.

**Comment by predatorian3** [comment]
Confirmed that this hack does work.

**Comment by AlmogCohen** [comment]
Same error here, same hack confirmed to be working...Are you sure this is not a Chef bug? @SteFletcher can you please link to the StackOverflow thread? I was unable to find it, might be helpful to fix this issue once and for all

**Comment by AlmogCohen** [comment]
Since I'm using RVM gemsets I had to make the patch suggested by @SteFletcher to every gemset, assuming knife was installed in gemset context

**Comment by coderanger** [comment]
Do any of you have encrypted private keys or maybe very old versions of OpenSSL?

**Comment by AlmogCohen** [comment]
Hmm:
* No private keys.
* My machine tends to be up to date (and is quite new anyway).
* `$ openssl version` gives `LibreSSL 2.2.7`. Which is kind of odd. But it seems that this is the version that apple bundled with newer versions of OSX.

**Comment by jaymzh** [comment]
Not sure why this is closed. I see the same issue on Linux trying to use a non-encrpted SSL key.

**Comment by coderanger** [comment]
@jaymzh What version of OpenSSL are you using? This appears to be related to older OpenSSL compared to both what I use on my laptop and what we ship in our installers.

**Comment by jaymzh** [comment]
rvm 2.4.1 which is where I think the relevant stuff comes from... openssl itself on my machine is quite new:

```
[phil@rider ~]$ openssl version
OpenSSL 1.1.0h  27 Mar 2018
```

The command at hand was `knife cookbook upload`, from taste-tester:

```
---- Begin output of knife cookbook upload scale_apache scale_mailman scale_reg -c /home/phil/.chef/knife-phil-taste-tester.rb ----
STDOUT: 
STDERR: ERROR: OpenSSL::OpenSSLError: password must be at least 4 bytes
```

Not much extra info with lots of verbosity:

```
[phil@rider (letsencrypt) scale-chef]$ knife cookbook upload scale_apache scale_mailman scale_reg -c /home/phil/.chef/knife-phil-taste-tester.rb -VVVV
INFO: Using configuration from /home/phil/.chef/knife-phil-taste-tester.rb
TRACE: No chefignore file found at /home/phil/build/git/scale-chef/chefignore no files will be ignored
ERROR: OpenSSL::OpenSSLError: password must be at least 4 bytes
[phil@rider (letsencrypt) scale-chef]$ knife --version
Chef: 14.3.37
```

**Comment by coderanger** [comment]
Okay, let me try to reproduce :)

**Comment by jadon1979** [comment]
OSX: LibreSSL 2.2.7
Chef DK: 3.1.0
Chef-client: 14.2.0

I had the exact same issue after updating chef.  Modifying line 95, as suggested, worked.

**Comment by kamaradclimber** [comment]
Same here, suddenly started to fail:  https://travis-ci.org/criteo-cookbooks/choregraphie/builds/423010172.

**Comment by cvasq** [comment]
I had this same issue today, updating the openssl gem fixed it for me.

`gem update openssl`

**Comment by jaymzh** [comment]
@coderanger and/o @lamont-granquist  this is still broken, afaict

**Comment by lamont-granquist** [comment]
So first of all RVM and ChefDK can't be used together.  You have to `rvm use system` to git rid of rvm and use the /opt/chefdk/bin path.

And this appears to be a problem with the openssl gem which shipped in the ruby 2.4.0 version used on rvm in travis.

This is not chef's problem.

People using travis need to realize that setting a "2.4" ruby version in travis does not roll forwards so users need to be explicitly using "2.4.9" as a version pin.

The alternative solution of pinning the openssl version in the Gemfile is fine as well, but this is not a problem with chef-client, this is a problem with the base ruby you are using and needs to be managed there.

**Comment by jaymzh** [comment]
@lamont-granquist why not depend on a minimum version of the openssl gem?
